src/main/java/org/myrobotlab/service/RasPi.java

```diff
@@ -376,6 +541,110 @@ public PinData publishPin(PinData pinData) {
     return pinData;
   }
 
+  public Map<Integer, Set<String>> publishScan() {
+    return validI2CAddresses;
+  }
+
+  public void read() {
+    log.debug("read task invoked");
+    List<PinData> pinArray = new ArrayList<>();
+    // load pin array
+    for (String pin : pinIndex.keySet()) {
+      PinDefinition pindef = pinIndex.get(pin);
+      if (pindef.isEnabled()) {
+        log.debug("pin {} enabled {}", pin, pindef.isEnabled());
+        int value = read(pin);
+        pindef.setValue(value);
+        PinData pd = new PinData(pin, value);
+        log.debug("pin data {}", pd);
+        pinArray.add(pd);
+      }
+    }
+
+    if (pinArray.size() > 0) {
+      PinData[] array = pinArray.toArray(new PinData[0]);
+      invoke("publishPinArray", new Object[] { array });
+    }
+  }
+
+  @Override
+  public int read(String pin) {
+
+    if (!pinIndex.containsKey(pin)) {
+      error("Pin %s not found", pin);
+      return -1;
+    }
+    PinDefinition pindef = pinIndex.get(pin);
+    GpioPinDigitalMultipurpose gpioPin = getGPIO(pin);
+    if (!gpioPin.isMode(PinMode.DIGITAL_INPUT)) {
+      pinMode(pin, "INPUT");
+    }
+    if (gpioPin.isLow()) {
+      pindef.setValue(0);
+      return 0;
+    } else {
+      pindef.setValue(1);
+      return 1;
+    }
+  }
+
+  @Override
+  public void reset() {
+    // TODO Auto-generated method stub
+    // reset pins/i2c devices/gpio pins
+  }
+
+  public void scan() {
+    scan(null);
+  }
+
+  public void scan(Integer busNumber) {
+
+    if (busNumber == null) {
+      busNumber = Integer.parseInt(bus);
+    }
+
+    try {
+
+      I2CBus bus = I2CFactory.getInstance(busNumber);
+
+      if (!validI2CAddresses.containsKey(busNumber)) {
+        validI2CAddresses.put(busNumber, new HashSet<>());
+      }
+
+      Set<String> addresses = validI2CAddresses.get(busNumber);
+
+      for (int i = 1; i < 128; i++) {

```

Comment:
Probably should put a comment explaining why i starts at 1

Reply:
some info says its reserved

Reserved Addresses:
0x00 (general call address)
0x01 to 0x07 (reserved)
0x78 to 0x7F (10-bit addressing, not usable in 7-bit addressing)

![image](https://github.com/MyRobotLab/myrobotlab/assets/1266144/449695a1-60cf-404d-b879-cf6dcd15af31)

some tools scan at 03 ... :shrug: 

it can also depend on if its 7-bit or 10-bit addressing (raspi is 7 I believe) 

I'd document it if I was confident I understood it.